### PR TITLE
Fix missing hashcode set in OrderedDictionary.SetAt

### LINKED ISSUE #103637: `OrderedDictionary.SetAt` bug
I encountered the following issue while adapting the new generic OrderedDictionary to STJ's JsonObject type:

```C#
[Fact]
public static void OrderedDictionary_SetAt_Bug()
{
    var dict = new OrderedDictionary<int, int>
    {
        [0] = 0
    };

    dict.SetAt(index: 0, key: 42, value: 42);

    Assert.Equal(1, dict.Count);
    Assert.Equal([new(42, 42)], dict);
    Assert.False(dict.ContainsKey(0));
    Assert.True(dict.ContainsKey(42)); // Fails
}
```

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/103637

### CODE REVIEW

File: src/libraries/System.Collections/tests/Generic/OrderedDictionary/OrderedDictionary.Generic.Tests.cs [hunk 1]
```diff
@@ -308,6 +308,25 @@ public void OrderedDictionary_Generic_SetAt_GetAt_Roundtrip(int count)
             }
         }
 
+        [Fact]
+        public void OrderedDictionary_SetAt_KeyValuePairSubsequentlyAvailable()
+        {
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/tests/Generic/OrderedDictionary/OrderedDictionary.Generic.Tests.cs:313]: Perhaps link back to the reporting issue?

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

File: src/libraries/System.Collections/tests/Generic/OrderedDictionary/OrderedDictionary.Generic.Tests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Collections/tests/Generic/OrderedDictionary/OrderedDictionary.Generic.Tests.cs:313]: I'd rather not. In a few weeks the fact that there was an issue won't matter, and this test won't be any different from every other test I added.

**Comment by stephentoub** [comment]
CI appears to be in a bad state

**Comment by stephentoub** [comment]
/ba-g ci is horked and this bug is blocking another p6 pr

File: src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs [hunk 2]
```diff
@@ -809,8 +809,8 @@ public void SetAt(int index, TKey key, TValue value)
             }
 
             // The key doesn't match that index. If it exists elsewhere in the collection, fail.
-            uint _ = 0, collisionCount = 0;
-            if (IndexOf(key, ref _, ref collisionCount) >= 0)
+            uint hashCode = 0, collisionCount = 0;
```
Comment (danmoseley) [inline src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:812]: Just curious why do we use uint for collision count? In general we use signed everywhere unless we need the full range.

Comment (stephentoub) [reply src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:812]: This just copies what `Dictionary<TKey, TValue>` does. I believe it uses uint just to make it very clear there's no possibility of overflow.

**Pull request merged**